# Required Imports

In [11]:
import os
import re
import random
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import numpy as np
from numpy import dot
from numpy.linalg import norm
from collections import defaultdict
from tqdm import tqdm

# -- NEW IMPORTS FOR GRAPHS --
from torch_geometric.data import Data, Dataset
from torch_geometric.loader import DataLoader
from torch_geometric.nn import GATConv, global_mean_pool, global_max_pool

## Parameters

In [12]:
# Configurazione PyG
RANDOM_SEED = 42
np.random.seed(RANDOM_SEED)
torch.manual_seed(RANDOM_SEED)

if torch.cuda.is_available():
    device = torch.device("cuda")
else:
    device = torch.device("cpu")
print(f'device being used: {device}')

# Configuration settings
n_classes = 40 
n_epochs = 100
train_batch_size = 32
test_batch_size = 1
learning_rate = 0.0005
num_node_features = 100 # Sostituisce img_w e img_h
num_nodes = 52

device being used: cpu


## Load Probe and Gallery sets

In [13]:
def extract_id_from_filename(filename):
    # Regex super resiliente: cerca un numero seguito da 'm' o 'f', ignorando se c'è un underscore o lettere extra
    match = re.search(r'(\d+[mf])', filename) 
    return match.group(1) if match else None

class GraphReIDTrainDataset(Dataset):
    def __init__(self, data_root):
        super().__init__()
        self.data_root = data_root
        self.samples, self.id_to_class = self._load_data()
        print(f"TRAIN: Trovati {len(self.samples)} grafi in {data_root}")

    def _load_data(self):
        paths = []
        for dirpath, _, filenames in os.walk(self.data_root):
            for f in filenames:
                if f.endswith('.pt'):
                    paths.append(os.path.join(dirpath, f))
        
        unique_ids = sorted(list(set([extract_id_from_filename(p) for p in paths if extract_id_from_filename(p)])))
        id_to_class = {orig_id: idx for idx, orig_id in enumerate(unique_ids)}
        
        samples = []
        for p in paths:
            pid = extract_id_from_filename(p)
            if pid:
                samples.append((p, id_to_class[pid]))
        return samples, id_to_class

    def len(self):
        return len(self.samples)

    def get(self, idx):
        path, target = self.samples[idx]
        data = torch.load(path, weights_only=False)
        data.y = torch.tensor([target], dtype=torch.long)
        data.path = path # Salviamo il path per il PKSampler
        return data

# PKSampler invariato
class PKSampler(torch.utils.data.Sampler):
    def __init__(self, dataset, P=8, K=4):
        self.dataset = dataset
        self.P = P
        self.K = K
        self.batch_size = self.P * self.K
        
        self.id_to_indices = defaultdict(list)
        for idx in range(len(self.dataset)):
            target = self.dataset.samples[idx][1]
            self.id_to_indices[target].append(idx)
        self.unique_ids = list(self.id_to_indices.keys())

    def __iter__(self):
        random.shuffle(self.unique_ids)
        batches = []
        for i in range(0, len(self.unique_ids), self.P):
            p_ids = self.unique_ids[i:i+self.P]
            if len(p_ids) < self.P: break
            batch = []
            for pid in p_ids:
                indices = self.id_to_indices[pid]
                if len(indices) >= self.K:
                    batch.extend(random.sample(indices, self.K))
                else:
                    batch.extend(np.random.choice(indices, size=self.K, replace=True))
            batches.append(batch)
        
        random.shuffle(batches)
        for batch in batches:
            yield from batch

    def __len__(self):
        return (len(self.unique_ids) // self.P) * self.batch_size

class GraphProbeGalleryValTest(Dataset):
    def __init__(self, data_root):
        super().__init__()
        self.data_root = data_root  
        self.data = self.read_folder()  
        self.samples = self.create_samples()  
        print(f"TEST: Trovati {len(self.data)} file totali. Generati {len(self.samples)} campioni Probe-Gallery.")

    def len(self):
        return len(self.samples)  

    def get(self, idx):
        probe_pth, gallery_pths, is_match, probe_target, gallery_targets = self.samples[idx]  
        probe_data = torch.load(probe_pth, weights_only=False)
        
        gallery_data_list = []
        for item in gallery_pths:
            gallery_data_list.append(torch.load(item, weights_only=False))
            
        return probe_data, gallery_data_list, is_match, probe_pth, gallery_pths, probe_target, gallery_targets

    def read_folder(self):
        paths = []  
        for dirpath, _, filenames in os.walk(self.data_root):
            files = [f for f in filenames if f.endswith('.pt')]
            # Metodo OS-indipendente per prendere l'ultima cartella (es. "Probe")
            label = os.path.basename(os.path.normpath(dirpath))
            for item in files:
                paths.append([os.path.join(dirpath, item), label])  
        return paths

    def create_samples(self):
        # Case insensitive per massima sicurezza
        probes = [x[0] for x in self.data if x[1].lower() == 'probe']  
        gallery = [x[0] for x in self.data if x[1].lower() == 'gallery']  
        samples = []
        for probe in probes:
            probe_id = extract_id_from_filename(probe)  
            if not probe_id: continue
            is_match, gallery_targets = [], []
            for item in gallery:
                item_id = extract_id_from_filename(item)  
                if not item_id: continue
                is_match.append(1 if probe_id == item_id else -1)
                gallery_targets.append(item_id)
            
            # Aggiunge alla lista solo se ha effettivamente trovato la gallery
            if len(gallery_targets) > 0:
                samples.append([probe, gallery, is_match, probe_id, gallery_targets])
        return samples

# Dataloaders Inizializzati
train_dataset = GraphReIDTrainDataset('./output/GraphData/D1/')
pk_sampler = PKSampler(train_dataset, P=8, K=4) 
train_dataloader = DataLoader(train_dataset, batch_size=32, sampler=pk_sampler, drop_last=True)

test_dataset = GraphProbeGalleryValTest('./output/GraphData/D2t/')
test_dataloader = DataLoader(test_dataset, batch_size=1, shuffle=False)

TRAIN: Trovati 800 grafi in ./output/GraphData/D1/
TEST: Trovati 820 file totali. Generati 410 campioni Probe-Gallery.


# Graph Attention Network

In [14]:
class GraphAttentionNetwork(nn.Module):
    def __init__(self, in_channels, temporal_dim, hidden_channels, out_channels, heads=4):
        super().__init__()
        
        # 1. Frequency Positional Encoding
        self.pos_emb = nn.Embedding(52, 16) 
        
        # 2A. High-Frequency Branch (Original small kernels)
        self.branch_short = nn.Sequential(
            nn.Conv1d(in_channels=1, out_channels=8, kernel_size=5, padding=2),
            nn.BatchNorm1d(8),
            nn.ReLU(),
            nn.MaxPool1d(2),
            nn.Conv1d(in_channels=8, out_channels=16, kernel_size=3, padding=1),
            nn.BatchNorm1d(16),
            nn.ReLU(),
            nn.AdaptiveAvgPool1d(1) 
        )
        
        # 2B. Macro-Movement Branch (Wide dilated kernels)
        self.branch_long = nn.Sequential(
            nn.Conv1d(in_channels=1, out_channels=8, kernel_size=15, padding=7),
            nn.BatchNorm1d(8),
            nn.ReLU(),
            nn.MaxPool1d(2),
            nn.Conv1d(in_channels=8, out_channels=16, kernel_size=9, padding=8, dilation=2),
            nn.BatchNorm1d(16),
            nn.ReLU(),
            nn.AdaptiveAvgPool1d(1) 
        )
        
        # Fusion layer: Combines Short (16) + Long (16) + Position (16) -> temporal_dim
        self.feature_fusion = nn.Sequential(
            nn.Linear(16 + 16 + 16, temporal_dim),
            nn.ReLU()
        )
        
        # 3. Graph Attention Layers
        self.conv1 = GATConv(temporal_dim, hidden_channels, heads=heads, dropout=0.3)
        self.conv2 = GATConv(hidden_channels * heads, out_channels, heads=1, concat=False, dropout=0.3)
        
    def forward(self, x, edge_index, batch):
        N = x.size(0) 
        
        # --- A. Inject Frequency Positional Encoding ---
        subcarrier_idx = (torch.arange(N, device=x.device) % 52)
        pos_features = self.pos_emb(subcarrier_idx) 
        
        # --- B. Extract Parallel Temporal Features ---
        x_seq = x.unsqueeze(1) 
        
        feat_short = self.branch_short(x_seq).squeeze(-1) # [N, 16]
        feat_long = self.branch_long(x_seq).squeeze(-1)   # [N, 16]
        
        # --- C. Feature Fusion ---
        node_features = torch.cat([feat_short, feat_long, pos_features], dim=1) # [N, 48]
        node_features = self.feature_fusion(node_features) 
        
        # --- D. Spatial Routing (GAT) ---
        x_g, alpha1 = self.conv1(node_features, edge_index, return_attention_weights=True)
        x_g = F.elu(x_g)
        
        x_g, alpha2 = self.conv2(x_g, edge_index, return_attention_weights=True)
        
        # --- E. Multi-Faceted Readout ---
        x_mean = global_mean_pool(x_g, batch)
        x_max = global_max_pool(x_g, batch)
        
        global_features = torch.cat([x_mean, x_max], dim=1)
        global_features = F.normalize(global_features, p=2, dim=1)
        
        return global_features, alpha2

# Graph Consistency Loss & Hard Triplet Loss

In [15]:
class GraphConsistencyTripletLoss(nn.Module):
    # Drop lambda_gc to 0.005 so it doesn't overpower the Triplet Loss
    def __init__(self, margin=0.3, lambda_gc=0.005): 
        super().__init__()
        self.ranking_loss = nn.MarginRankingLoss(margin=margin)
        self.lambda_gc = lambda_gc

    def forward(self, embeddings, attentions, targets, batch):
        # --- 1. Hard Triplet Loss Classica ---
        n = embeddings.size(0)
        dist = torch.pow(embeddings, 2).sum(dim=1, keepdim=True).expand(n, n)
        dist = dist + dist.t()
        dist.addmm_(embeddings, embeddings.t(), beta=1, alpha=-2)
        dist = dist.clamp(min=1e-12).sqrt()

        mask = targets.expand(n, n).eq(targets.expand(n, n).t())
        dist_ap, dist_an = [], []

        for i in range(n):
            dist_ap.append(dist[i][mask[i]].max().unsqueeze(0))
            dist_an.append(dist[i][mask[i] == 0].min().unsqueeze(0))

        dist_ap = torch.cat(dist_ap)
        dist_an = torch.cat(dist_an)
        y = torch.ones_like(dist_an)
        triplet_loss = self.ranking_loss(dist_an, dist_ap, y)
        
        # --- 2. Graph Consistency Loss (L2 Regularization) ---
        edge_index, alpha = attentions 
        
        # Taking the mean of the squared attention weights penalizes 
        # the model if it tries to collapse all attention onto a single edge.
        gc_loss = torch.mean(alpha ** 2)
        
        return triplet_loss + (self.lambda_gc * gc_loss)

# Training and Testing functions

In [16]:
def compute_ap(pos_ranks):
    """Calcolo corretto della Average Precision"""
    if len(pos_ranks) == 0:
        return 0.0
    ap = 0.0
    for i, pos in enumerate(pos_ranks):
        ap += (i + 1) / (pos + 1)
    return ap / len(pos_ranks)


from torch_geometric.data import Batch

def testing_step(model, test_dataloader):
    model.eval()

    test_dataset = test_dataloader.dataset if hasattr(test_dataloader, 'dataset') else test_dataloader

    if len(test_dataset.samples) == 0:
        print("\nWARNING: Test Dataset è vuoto!")
        return 0.0, 0.0, 0.0, 0.0

    # 1. ESTRAZIONE DELLA GALLERY (UNA VOLTA SOLA)
    # Prendiamo i percorsi della gallery dal primissimo campione (sono uguali per tutti)
    gallery_pths = test_dataset.samples[0][1] 
    
    print(f"\nEstrazione feature della Gallery ({len(gallery_pths)} grafi)...")
    gallery_features_list = []
    gallery_ids = []

    with torch.no_grad():
        # Processiamo la gallery a blocchi di 32 per non saturare la RAM
        batch_size = 32
        for i in range(0, len(gallery_pths), batch_size):
            batch_paths = gallery_pths[i:i+batch_size]
            batch_graphs = [torch.load(p, weights_only=False) for p in batch_paths]
            g_batch = Batch.from_data_list(batch_graphs).to(device)

            # Estrai le feature e spostale su CPU per tenerle in memoria
            features, _ = model(g_batch.x, g_batch.edge_index, g_batch.batch)
            gallery_features_list.append(features.cpu())

            # Salva gli ID nativi
            batch_ids = g_batch.subject_id
            if isinstance(batch_ids, str): 
                batch_ids = [batch_ids]
            gallery_ids.extend(batch_ids)

        gallery_features = torch.cat(gallery_features_list, dim=0)

    # 2. VALUTAZIONE DELLE PROBE
    print("\nValutazione Probes in corso...")
    samples = 0  
    K = 10
    correct_at_k = np.zeros(K)
    AP_scores = []  
    rank_matrix = []

    with torch.no_grad():
        # Invece di usare il DataLoader, iteriamo direttamente sulla lista dei campioni
        for sample in tqdm(test_dataset.samples, leave=False):
            probe_pth = sample[0] # Il percorso della probe è l'elemento 0
            
            # Carichiamo solo la singola probe (1 lettura da disco invece di 400+)
            probe_data = torch.load(probe_pth, weights_only=False)
            probe_batch = Batch.from_data_list([probe_data]).to(device)

            # Feature della probe
            probe_feat, _ = model(probe_batch.x, probe_batch.edge_index, probe_batch.batch)
            probe_feat = probe_feat.cpu()

            scores = []
            
            # Optimized vectorized code
            import torch.nn.functional as F
            
            # probe_feat is shape [1, feature_dim]
            # gallery_features is shape [num_gallery, feature_dim]
            # This computes cosine similarity for the entire gallery in one fast tensor operation
            similarities = F.cosine_similarity(probe_feat, gallery_features)
            
            # Zip the results back into the expected scores list format
            scores = [[sim.item(), gal_id] for sim, gal_id in zip(similarities, gallery_ids)]
            
            samples += 1
            topk = sorted(scores, key=lambda element: (element[0]), reverse=True)
            
            p_id = probe_batch.subject_id[0] if isinstance(probe_batch.subject_id, list) else probe_batch.subject_id
            rank_matrix.append([p_id, topk])

            # -- CALCOLO mAP --
            pos_ranks = []
            for i, score in enumerate(topk):
                gal_id = score[1]
                if p_id == gal_id:
                    pos_ranks.append(i)
                    
            if pos_ranks:
                AP = compute_ap(pos_ranks)  
                AP_scores.append(AP)  

        # -- CALCOLO MATRICE DEI RANK --
        for idx, probe in enumerate(rank_matrix):
            p_id = probe[0]
            for i in range(K):
                gal_id = probe[1][i][1]
                if p_id == gal_id:
                    # Increment this rank and all subsequent ranks by 1
                    correct_at_k[i:] += 1
                    break

    # Convert matches to percentages
    rank_accs = (correct_at_k / samples) * 100.0
    mAP = np.mean(AP_scores) if AP_scores else 0.0

    print(f'Total samples: {samples}, Rank #1: {rank_accs[0]:.3f}%, Rank #2: {rank_accs[1]:.3f}%, Rank #3: {rank_accs[2]:.3f}%, Rank #4: {rank_accs[3]:.3f}%, Rank #5: {rank_accs[4]:.3f}%, Rank #6: {rank_accs[5]:.3f}%, Rank #7: {rank_accs[6]:.3f}%, Rank #8: {rank_accs[7]:.3f}%, Rank #9: {rank_accs[8]:.3f}%, Rank #10: {rank_accs[9]:.3f}%, mAP:{mAP*100.0:.3f}%')
    
    # Return Rank 1, Rank 5, Rank 10 as decimals
    return correct_at_k[0]/samples, correct_at_k[4]/samples, correct_at_k[9]/samples, mAP

In [17]:
def training_model(model, train_dataloader, test_dataloader, optimizer, criterion, n_epochs):
    bestRank1 = 0.0

    for epoch in range(n_epochs):
        model.train()
        train_loss = 0.0

        for batch_data in tqdm(train_dataloader, leave=False, total=len(train_dataloader)):
            # I dati PyG includono già x, edge_index, batch e y(targets)
            batch_data = batch_data.to(device)
            targets = batch_data.y.view(-1)
            
            # Forward sulla GAT
            features, attentions = model(batch_data.x, batch_data.edge_index, batch_data.batch)
            
            # Loss combinata
            loss = criterion(features, attentions, targets, batch_data.batch)

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()
            train_loss += loss.item()

        # Validation step
        rank1, rank5, rank10, _ = testing_step(model, test_dataloader)

        # Salvataggi
        models_folder = os.path.join('./output/models/')
        os.makedirs(models_folder, exist_ok=True)
        if rank1 > bestRank1:
            torch.save(model.state_dict(), f'{models_folder}/best_GNN_state_r1.bin')
            bestRank1 = rank1
            
        print(f"Epoch: [{epoch+1}/{n_epochs}]| Loss: {train_loss/len(train_dataloader):.5f}")

## Model Parameters

In [18]:
# Initialize the upgraded GNN model (temporal_dim=128 maps the 100 packets into a richer space)
model = GraphAttentionNetwork(in_channels=100, temporal_dim=128, hidden_channels=256, out_channels=256).to(device)
criterion = GraphConsistencyTripletLoss()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)

## Network Training

In [19]:
# Start training model
training_model(model, train_dataloader, test_dataloader, optimizer, criterion, n_epochs)


Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 48.293%, Rank #2: 60.000%, Rank #3: 65.854%, Rank #4: 69.268%, Rank #5: 72.683%, Rank #6: 75.854%, Rank #7: 77.073%, Rank #8: 78.293%, Rank #9: 79.756%, Rank #10: 81.220%, mAP:25.049%
Epoch: [1/100]| Loss: 0.32902



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 53.415%, Rank #2: 66.341%, Rank #3: 71.951%, Rank #4: 75.366%, Rank #5: 78.049%, Rank #6: 78.780%, Rank #7: 80.244%, Rank #8: 81.463%, Rank #9: 82.927%, Rank #10: 83.659%, mAP:28.429%
Epoch: [2/100]| Loss: 0.31745



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 56.098%, Rank #2: 67.805%, Rank #3: 74.390%, Rank #4: 78.049%, Rank #5: 81.463%, Rank #6: 82.927%, Rank #7: 84.390%, Rank #8: 85.122%, Rank #9: 85.122%, Rank #10: 85.854%, mAP:34.613%
Epoch: [3/100]| Loss: 0.31231



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 62.927%, Rank #2: 74.878%, Rank #3: 79.756%, Rank #4: 81.463%, Rank #5: 83.415%, Rank #6: 84.878%, Rank #7: 85.610%, Rank #8: 87.317%, Rank #9: 88.293%, Rank #10: 88.293%, mAP:40.595%
Epoch: [4/100]| Loss: 0.31133



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 62.439%, Rank #2: 75.610%, Rank #3: 82.195%, Rank #4: 83.659%, Rank #5: 85.854%, Rank #6: 87.073%, Rank #7: 88.537%, Rank #8: 89.024%, Rank #9: 89.756%, Rank #10: 90.000%, mAP:44.733%
Epoch: [5/100]| Loss: 0.30599



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 63.902%, Rank #2: 76.829%, Rank #3: 82.683%, Rank #4: 84.634%, Rank #5: 86.341%, Rank #6: 87.805%, Rank #7: 89.268%, Rank #8: 90.244%, Rank #9: 91.220%, Rank #10: 91.707%, mAP:47.604%
Epoch: [6/100]| Loss: 0.30727



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 66.585%, Rank #2: 77.317%, Rank #3: 80.976%, Rank #4: 84.146%, Rank #5: 86.829%, Rank #6: 88.780%, Rank #7: 89.268%, Rank #8: 90.732%, Rank #9: 92.195%, Rank #10: 92.683%, mAP:50.060%
Epoch: [7/100]| Loss: 0.30649



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 64.634%, Rank #2: 74.878%, Rank #3: 80.000%, Rank #4: 82.683%, Rank #5: 85.122%, Rank #6: 86.829%, Rank #7: 87.561%, Rank #8: 89.512%, Rank #9: 90.244%, Rank #10: 91.220%, mAP:48.749%
Epoch: [8/100]| Loss: 0.30058



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 63.902%, Rank #2: 77.317%, Rank #3: 80.976%, Rank #4: 83.415%, Rank #5: 84.390%, Rank #6: 86.829%, Rank #7: 87.561%, Rank #8: 88.049%, Rank #9: 89.268%, Rank #10: 90.732%, mAP:47.348%
Epoch: [9/100]| Loss: 0.30860



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 63.171%, Rank #2: 75.122%, Rank #3: 80.976%, Rank #4: 83.171%, Rank #5: 85.366%, Rank #6: 86.341%, Rank #7: 89.024%, Rank #8: 90.732%, Rank #9: 91.707%, Rank #10: 92.195%, mAP:49.028%
Epoch: [10/100]| Loss: 0.30445



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 62.927%, Rank #2: 73.659%, Rank #3: 81.463%, Rank #4: 83.659%, Rank #5: 86.341%, Rank #6: 88.049%, Rank #7: 89.512%, Rank #8: 90.488%, Rank #9: 91.707%, Rank #10: 92.195%, mAP:48.922%
Epoch: [11/100]| Loss: 0.29557



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 65.122%, Rank #2: 78.049%, Rank #3: 82.683%, Rank #4: 84.878%, Rank #5: 85.854%, Rank #6: 87.561%, Rank #7: 90.244%, Rank #8: 91.220%, Rank #9: 91.463%, Rank #10: 91.951%, mAP:50.248%
Epoch: [12/100]| Loss: 0.29611



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 67.561%, Rank #2: 79.268%, Rank #3: 83.171%, Rank #4: 85.854%, Rank #5: 88.049%, Rank #6: 88.780%, Rank #7: 90.488%, Rank #8: 91.220%, Rank #9: 92.439%, Rank #10: 92.927%, mAP:51.693%
Epoch: [13/100]| Loss: 0.29269



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 65.610%, Rank #2: 76.098%, Rank #3: 84.390%, Rank #4: 88.293%, Rank #5: 89.512%, Rank #6: 90.732%, Rank #7: 90.976%, Rank #8: 91.463%, Rank #9: 91.707%, Rank #10: 92.195%, mAP:53.102%
Epoch: [14/100]| Loss: 0.27373



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 67.073%, Rank #2: 79.024%, Rank #3: 83.171%, Rank #4: 85.610%, Rank #5: 87.805%, Rank #6: 89.512%, Rank #7: 90.488%, Rank #8: 91.707%, Rank #9: 92.439%, Rank #10: 92.927%, mAP:52.088%
Epoch: [15/100]| Loss: 0.30228



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 67.561%, Rank #2: 79.512%, Rank #3: 83.415%, Rank #4: 85.610%, Rank #5: 88.293%, Rank #6: 89.024%, Rank #7: 89.512%, Rank #8: 92.439%, Rank #9: 92.683%, Rank #10: 92.683%, mAP:52.481%
Epoch: [16/100]| Loss: 0.28859



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 69.268%, Rank #2: 78.293%, Rank #3: 82.683%, Rank #4: 86.098%, Rank #5: 88.049%, Rank #6: 88.293%, Rank #7: 89.512%, Rank #8: 90.976%, Rank #9: 91.707%, Rank #10: 92.195%, mAP:52.996%
Epoch: [17/100]| Loss: 0.28463



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 70.732%, Rank #2: 80.244%, Rank #3: 83.415%, Rank #4: 87.073%, Rank #5: 88.293%, Rank #6: 89.024%, Rank #7: 90.000%, Rank #8: 91.707%, Rank #9: 91.951%, Rank #10: 92.195%, mAP:56.230%
Epoch: [18/100]| Loss: 0.29548



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 70.488%, Rank #2: 79.024%, Rank #3: 83.171%, Rank #4: 85.854%, Rank #5: 88.049%, Rank #6: 89.024%, Rank #7: 90.488%, Rank #8: 91.463%, Rank #9: 92.439%, Rank #10: 93.171%, mAP:54.733%
Epoch: [19/100]| Loss: 0.28339



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 68.293%, Rank #2: 79.024%, Rank #3: 83.415%, Rank #4: 85.610%, Rank #5: 87.561%, Rank #6: 88.049%, Rank #7: 88.780%, Rank #8: 89.756%, Rank #9: 90.488%, Rank #10: 91.220%, mAP:54.020%
Epoch: [20/100]| Loss: 0.27506



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 66.829%, Rank #2: 77.561%, Rank #3: 82.683%, Rank #4: 84.878%, Rank #5: 86.341%, Rank #6: 87.073%, Rank #7: 88.293%, Rank #8: 89.512%, Rank #9: 89.512%, Rank #10: 90.732%, mAP:53.275%
Epoch: [21/100]| Loss: 0.27420



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 66.098%, Rank #2: 77.317%, Rank #3: 80.244%, Rank #4: 83.415%, Rank #5: 85.122%, Rank #6: 86.341%, Rank #7: 88.049%, Rank #8: 88.537%, Rank #9: 89.024%, Rank #10: 89.756%, mAP:51.806%
Epoch: [22/100]| Loss: 0.24986



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 67.317%, Rank #2: 76.829%, Rank #3: 80.732%, Rank #4: 82.927%, Rank #5: 85.610%, Rank #6: 87.561%, Rank #7: 89.756%, Rank #8: 90.732%, Rank #9: 91.951%, Rank #10: 92.195%, mAP:52.203%
Epoch: [23/100]| Loss: 0.28228



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 67.317%, Rank #2: 78.049%, Rank #3: 82.439%, Rank #4: 85.610%, Rank #5: 87.317%, Rank #6: 88.293%, Rank #7: 88.780%, Rank #8: 89.756%, Rank #9: 90.976%, Rank #10: 92.195%, mAP:53.464%
Epoch: [24/100]| Loss: 0.25781



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 68.293%, Rank #2: 77.805%, Rank #3: 81.951%, Rank #4: 85.366%, Rank #5: 88.293%, Rank #6: 89.756%, Rank #7: 90.976%, Rank #8: 91.707%, Rank #9: 92.195%, Rank #10: 92.927%, mAP:55.626%
Epoch: [25/100]| Loss: 0.23445



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 67.073%, Rank #2: 76.829%, Rank #3: 81.220%, Rank #4: 83.902%, Rank #5: 87.073%, Rank #6: 87.805%, Rank #7: 88.780%, Rank #8: 90.732%, Rank #9: 92.439%, Rank #10: 92.927%, mAP:54.554%
Epoch: [26/100]| Loss: 0.20964



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 70.244%, Rank #2: 77.805%, Rank #3: 82.195%, Rank #4: 85.854%, Rank #5: 88.293%, Rank #6: 89.268%, Rank #7: 90.000%, Rank #8: 90.976%, Rank #9: 91.707%, Rank #10: 92.927%, mAP:55.155%
Epoch: [27/100]| Loss: 0.25239



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 70.244%, Rank #2: 80.488%, Rank #3: 85.122%, Rank #4: 87.561%, Rank #5: 89.268%, Rank #6: 90.488%, Rank #7: 90.976%, Rank #8: 91.463%, Rank #9: 92.195%, Rank #10: 92.683%, mAP:56.549%
Epoch: [28/100]| Loss: 0.24444



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 69.268%, Rank #2: 78.780%, Rank #3: 86.098%, Rank #4: 88.537%, Rank #5: 89.756%, Rank #6: 90.244%, Rank #7: 91.220%, Rank #8: 91.951%, Rank #9: 92.195%, Rank #10: 93.171%, mAP:55.895%
Epoch: [29/100]| Loss: 0.24280



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 67.073%, Rank #2: 76.341%, Rank #3: 81.707%, Rank #4: 86.098%, Rank #5: 87.317%, Rank #6: 88.780%, Rank #7: 90.000%, Rank #8: 90.488%, Rank #9: 91.220%, Rank #10: 92.439%, mAP:53.675%
Epoch: [30/100]| Loss: 0.30265



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.902%, Rank #2: 81.463%, Rank #3: 84.634%, Rank #4: 87.317%, Rank #5: 89.512%, Rank #6: 90.976%, Rank #7: 91.707%, Rank #8: 92.927%, Rank #9: 93.659%, Rank #10: 93.659%, mAP:57.823%
Epoch: [31/100]| Loss: 0.24208



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.659%, Rank #2: 81.220%, Rank #3: 84.146%, Rank #4: 87.561%, Rank #5: 88.537%, Rank #6: 91.220%, Rank #7: 92.195%, Rank #8: 93.171%, Rank #9: 93.902%, Rank #10: 94.390%, mAP:59.394%
Epoch: [32/100]| Loss: 0.26051



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.927%, Rank #2: 81.707%, Rank #3: 86.098%, Rank #4: 88.537%, Rank #5: 90.488%, Rank #6: 92.439%, Rank #7: 92.683%, Rank #8: 93.902%, Rank #9: 94.390%, Rank #10: 95.366%, mAP:62.108%
Epoch: [33/100]| Loss: 0.21417



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.610%, Rank #2: 80.000%, Rank #3: 85.122%, Rank #4: 87.805%, Rank #5: 90.244%, Rank #6: 90.976%, Rank #7: 91.707%, Rank #8: 93.171%, Rank #9: 94.146%, Rank #10: 94.634%, mAP:60.561%
Epoch: [34/100]| Loss: 0.25452



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.683%, Rank #2: 79.512%, Rank #3: 84.878%, Rank #4: 87.317%, Rank #5: 88.780%, Rank #6: 90.488%, Rank #7: 91.951%, Rank #8: 92.683%, Rank #9: 93.659%, Rank #10: 93.902%, mAP:59.597%
Epoch: [35/100]| Loss: 0.23150



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 71.951%, Rank #2: 79.512%, Rank #3: 83.171%, Rank #4: 86.098%, Rank #5: 88.293%, Rank #6: 90.000%, Rank #7: 91.463%, Rank #8: 91.707%, Rank #9: 92.195%, Rank #10: 93.415%, mAP:59.520%
Epoch: [36/100]| Loss: 0.21867



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 69.512%, Rank #2: 77.317%, Rank #3: 81.951%, Rank #4: 84.634%, Rank #5: 86.585%, Rank #6: 88.293%, Rank #7: 89.512%, Rank #8: 90.488%, Rank #9: 90.732%, Rank #10: 91.220%, mAP:55.991%
Epoch: [37/100]| Loss: 0.24032



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 70.244%, Rank #2: 80.244%, Rank #3: 84.146%, Rank #4: 86.341%, Rank #5: 88.537%, Rank #6: 89.268%, Rank #7: 90.000%, Rank #8: 91.220%, Rank #9: 92.439%, Rank #10: 93.902%, mAP:56.483%
Epoch: [38/100]| Loss: 0.21261



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.366%, Rank #2: 82.439%, Rank #3: 85.366%, Rank #4: 87.073%, Rank #5: 90.000%, Rank #6: 90.976%, Rank #7: 93.171%, Rank #8: 93.659%, Rank #9: 95.122%, Rank #10: 95.122%, mAP:61.102%
Epoch: [39/100]| Loss: 0.25944



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.634%, Rank #2: 81.463%, Rank #3: 87.317%, Rank #4: 88.537%, Rank #5: 91.463%, Rank #6: 92.439%, Rank #7: 93.171%, Rank #8: 94.146%, Rank #9: 94.634%, Rank #10: 94.878%, mAP:62.127%
Epoch: [40/100]| Loss: 0.23487



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.878%, Rank #2: 81.951%, Rank #3: 87.805%, Rank #4: 90.000%, Rank #5: 93.171%, Rank #6: 93.415%, Rank #7: 94.390%, Rank #8: 94.878%, Rank #9: 94.878%, Rank #10: 95.122%, mAP:62.055%
Epoch: [41/100]| Loss: 0.20697



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.610%, Rank #2: 83.415%, Rank #3: 87.805%, Rank #4: 89.756%, Rank #5: 91.463%, Rank #6: 93.171%, Rank #7: 93.902%, Rank #8: 94.634%, Rank #9: 95.122%, Rank #10: 95.366%, mAP:62.460%
Epoch: [42/100]| Loss: 0.18774



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.610%, Rank #2: 82.683%, Rank #3: 87.805%, Rank #4: 89.268%, Rank #5: 90.244%, Rank #6: 93.659%, Rank #7: 94.390%, Rank #8: 94.878%, Rank #9: 95.366%, Rank #10: 95.366%, mAP:61.441%
Epoch: [43/100]| Loss: 0.19949



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.195%, Rank #2: 80.488%, Rank #3: 84.878%, Rank #4: 86.341%, Rank #5: 88.293%, Rank #6: 90.000%, Rank #7: 90.732%, Rank #8: 92.927%, Rank #9: 93.902%, Rank #10: 94.390%, mAP:60.645%
Epoch: [44/100]| Loss: 0.18642



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.415%, Rank #2: 81.463%, Rank #3: 85.122%, Rank #4: 87.073%, Rank #5: 88.780%, Rank #6: 90.976%, Rank #7: 91.220%, Rank #8: 91.707%, Rank #9: 92.195%, Rank #10: 92.683%, mAP:60.362%
Epoch: [45/100]| Loss: 0.21778



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.171%, Rank #2: 79.024%, Rank #3: 83.171%, Rank #4: 87.073%, Rank #5: 89.024%, Rank #6: 90.488%, Rank #7: 91.463%, Rank #8: 92.195%, Rank #9: 92.927%, Rank #10: 93.171%, mAP:61.288%
Epoch: [46/100]| Loss: 0.13801



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.415%, Rank #2: 78.780%, Rank #3: 83.902%, Rank #4: 85.854%, Rank #5: 88.537%, Rank #6: 90.488%, Rank #7: 90.976%, Rank #8: 91.707%, Rank #9: 92.927%, Rank #10: 93.415%, mAP:59.659%
Epoch: [47/100]| Loss: 0.20270



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.366%, Rank #2: 81.707%, Rank #3: 85.366%, Rank #4: 87.073%, Rank #5: 89.756%, Rank #6: 91.707%, Rank #7: 91.951%, Rank #8: 92.683%, Rank #9: 92.927%, Rank #10: 93.902%, mAP:61.886%
Epoch: [48/100]| Loss: 0.20883



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 70.000%, Rank #2: 82.439%, Rank #3: 85.610%, Rank #4: 89.024%, Rank #5: 91.220%, Rank #6: 92.195%, Rank #7: 92.439%, Rank #8: 92.927%, Rank #9: 93.415%, Rank #10: 94.146%, mAP:60.644%
Epoch: [49/100]| Loss: 0.19751



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.927%, Rank #2: 82.439%, Rank #3: 86.341%, Rank #4: 88.293%, Rank #5: 89.756%, Rank #6: 91.220%, Rank #7: 92.439%, Rank #8: 92.683%, Rank #9: 93.415%, Rank #10: 94.146%, mAP:60.173%
Epoch: [50/100]| Loss: 0.21036



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.927%, Rank #2: 82.927%, Rank #3: 86.341%, Rank #4: 89.512%, Rank #5: 90.732%, Rank #6: 91.220%, Rank #7: 92.439%, Rank #8: 93.415%, Rank #9: 93.659%, Rank #10: 94.634%, mAP:61.040%
Epoch: [51/100]| Loss: 0.18372



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.171%, Rank #2: 82.927%, Rank #3: 87.317%, Rank #4: 89.756%, Rank #5: 90.976%, Rank #6: 92.439%, Rank #7: 93.659%, Rank #8: 94.878%, Rank #9: 95.366%, Rank #10: 95.854%, mAP:62.116%
Epoch: [52/100]| Loss: 0.20744



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.927%, Rank #2: 80.976%, Rank #3: 86.585%, Rank #4: 88.780%, Rank #5: 90.000%, Rank #6: 92.439%, Rank #7: 93.659%, Rank #8: 93.902%, Rank #9: 95.122%, Rank #10: 95.854%, mAP:60.615%
Epoch: [53/100]| Loss: 0.21870



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.146%, Rank #2: 82.439%, Rank #3: 86.341%, Rank #4: 88.780%, Rank #5: 90.732%, Rank #6: 91.951%, Rank #7: 92.927%, Rank #8: 93.902%, Rank #9: 94.634%, Rank #10: 95.122%, mAP:60.879%
Epoch: [54/100]| Loss: 0.22556



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.634%, Rank #2: 82.195%, Rank #3: 87.073%, Rank #4: 90.976%, Rank #5: 92.195%, Rank #6: 93.171%, Rank #7: 94.634%, Rank #8: 95.610%, Rank #9: 95.854%, Rank #10: 96.098%, mAP:64.300%
Epoch: [55/100]| Loss: 0.20802



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.634%, Rank #2: 81.951%, Rank #3: 86.098%, Rank #4: 88.780%, Rank #5: 91.707%, Rank #6: 92.927%, Rank #7: 93.415%, Rank #8: 94.390%, Rank #9: 94.390%, Rank #10: 95.366%, mAP:61.805%
Epoch: [56/100]| Loss: 0.17874



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.634%, Rank #2: 81.951%, Rank #3: 85.610%, Rank #4: 86.341%, Rank #5: 89.512%, Rank #6: 90.976%, Rank #7: 92.439%, Rank #8: 92.683%, Rank #9: 93.659%, Rank #10: 94.390%, mAP:62.105%
Epoch: [57/100]| Loss: 0.19474



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 71.463%, Rank #2: 81.220%, Rank #3: 87.073%, Rank #4: 88.780%, Rank #5: 90.976%, Rank #6: 92.439%, Rank #7: 92.927%, Rank #8: 93.659%, Rank #9: 94.634%, Rank #10: 95.122%, mAP:62.639%
Epoch: [58/100]| Loss: 0.19544



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.634%, Rank #2: 83.171%, Rank #3: 86.098%, Rank #4: 88.293%, Rank #5: 91.463%, Rank #6: 92.683%, Rank #7: 93.659%, Rank #8: 94.390%, Rank #9: 94.878%, Rank #10: 95.366%, mAP:61.598%
Epoch: [59/100]| Loss: 0.17894



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.854%, Rank #2: 83.902%, Rank #3: 86.829%, Rank #4: 90.488%, Rank #5: 93.171%, Rank #6: 94.878%, Rank #7: 95.610%, Rank #8: 95.854%, Rank #9: 96.341%, Rank #10: 96.585%, mAP:62.500%
Epoch: [60/100]| Loss: 0.13247



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.878%, Rank #2: 85.366%, Rank #3: 88.537%, Rank #4: 90.000%, Rank #5: 91.951%, Rank #6: 92.683%, Rank #7: 93.415%, Rank #8: 95.366%, Rank #9: 96.098%, Rank #10: 96.585%, mAP:64.157%
Epoch: [61/100]| Loss: 0.19617



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.171%, Rank #2: 83.171%, Rank #3: 87.805%, Rank #4: 90.488%, Rank #5: 91.707%, Rank #6: 93.902%, Rank #7: 94.878%, Rank #8: 95.122%, Rank #9: 95.854%, Rank #10: 96.098%, mAP:65.369%
Epoch: [62/100]| Loss: 0.20339



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.902%, Rank #2: 81.463%, Rank #3: 84.146%, Rank #4: 87.073%, Rank #5: 88.537%, Rank #6: 90.976%, Rank #7: 92.195%, Rank #8: 93.659%, Rank #9: 93.902%, Rank #10: 94.390%, mAP:64.181%
Epoch: [63/100]| Loss: 0.17207



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.439%, Rank #2: 80.732%, Rank #3: 84.878%, Rank #4: 87.317%, Rank #5: 90.000%, Rank #6: 91.220%, Rank #7: 91.951%, Rank #8: 93.415%, Rank #9: 94.146%, Rank #10: 94.878%, mAP:63.553%
Epoch: [64/100]| Loss: 0.16211



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.854%, Rank #2: 82.439%, Rank #3: 86.341%, Rank #4: 89.024%, Rank #5: 92.195%, Rank #6: 93.659%, Rank #7: 94.634%, Rank #8: 95.854%, Rank #9: 96.098%, Rank #10: 96.341%, mAP:63.432%
Epoch: [65/100]| Loss: 0.19661



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 76.341%, Rank #2: 84.634%, Rank #3: 87.073%, Rank #4: 90.000%, Rank #5: 91.951%, Rank #6: 93.902%, Rank #7: 94.878%, Rank #8: 95.610%, Rank #9: 95.854%, Rank #10: 96.098%, mAP:64.428%
Epoch: [66/100]| Loss: 0.13882



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 76.098%, Rank #2: 83.659%, Rank #3: 87.073%, Rank #4: 89.268%, Rank #5: 90.732%, Rank #6: 91.707%, Rank #7: 93.415%, Rank #8: 95.366%, Rank #9: 95.854%, Rank #10: 96.098%, mAP:64.805%
Epoch: [67/100]| Loss: 0.15941



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 76.585%, Rank #2: 83.171%, Rank #3: 86.829%, Rank #4: 87.561%, Rank #5: 89.512%, Rank #6: 91.463%, Rank #7: 92.195%, Rank #8: 94.634%, Rank #9: 95.122%, Rank #10: 95.122%, mAP:64.606%
Epoch: [68/100]| Loss: 0.14512



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.146%, Rank #2: 80.976%, Rank #3: 83.659%, Rank #4: 85.610%, Rank #5: 88.780%, Rank #6: 90.732%, Rank #7: 92.195%, Rank #8: 93.171%, Rank #9: 94.878%, Rank #10: 95.122%, mAP:61.856%
Epoch: [69/100]| Loss: 0.17363



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.146%, Rank #2: 82.683%, Rank #3: 86.829%, Rank #4: 89.024%, Rank #5: 90.244%, Rank #6: 91.707%, Rank #7: 92.195%, Rank #8: 93.902%, Rank #9: 95.366%, Rank #10: 96.098%, mAP:62.150%
Epoch: [70/100]| Loss: 0.19655



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.146%, Rank #2: 82.927%, Rank #3: 86.829%, Rank #4: 90.000%, Rank #5: 90.976%, Rank #6: 92.195%, Rank #7: 93.902%, Rank #8: 94.634%, Rank #9: 94.878%, Rank #10: 95.610%, mAP:64.548%
Epoch: [71/100]| Loss: 0.15386



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 76.098%, Rank #2: 82.927%, Rank #3: 86.341%, Rank #4: 88.049%, Rank #5: 90.244%, Rank #6: 91.220%, Rank #7: 92.683%, Rank #8: 94.146%, Rank #9: 94.878%, Rank #10: 95.122%, mAP:64.077%
Epoch: [72/100]| Loss: 0.13993



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 76.098%, Rank #2: 82.683%, Rank #3: 86.341%, Rank #4: 88.780%, Rank #5: 90.976%, Rank #6: 92.927%, Rank #7: 94.146%, Rank #8: 94.390%, Rank #9: 94.878%, Rank #10: 95.610%, mAP:62.743%
Epoch: [73/100]| Loss: 0.14486



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.415%, Rank #2: 81.463%, Rank #3: 85.122%, Rank #4: 88.780%, Rank #5: 90.976%, Rank #6: 91.463%, Rank #7: 92.683%, Rank #8: 93.902%, Rank #9: 94.146%, Rank #10: 94.390%, mAP:60.888%
Epoch: [74/100]| Loss: 0.15059



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 76.341%, Rank #2: 82.195%, Rank #3: 87.805%, Rank #4: 90.000%, Rank #5: 91.707%, Rank #6: 92.439%, Rank #7: 93.171%, Rank #8: 94.390%, Rank #9: 94.878%, Rank #10: 95.854%, mAP:64.535%
Epoch: [75/100]| Loss: 0.19331



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 76.098%, Rank #2: 81.463%, Rank #3: 86.585%, Rank #4: 88.780%, Rank #5: 91.220%, Rank #6: 92.683%, Rank #7: 93.415%, Rank #8: 94.146%, Rank #9: 95.122%, Rank #10: 95.854%, mAP:64.345%
Epoch: [76/100]| Loss: 0.16900



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.390%, Rank #2: 82.683%, Rank #3: 86.341%, Rank #4: 88.049%, Rank #5: 90.244%, Rank #6: 92.683%, Rank #7: 94.390%, Rank #8: 95.122%, Rank #9: 95.366%, Rank #10: 96.341%, mAP:63.612%
Epoch: [77/100]| Loss: 0.13975



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.195%, Rank #2: 82.927%, Rank #3: 86.341%, Rank #4: 89.512%, Rank #5: 91.463%, Rank #6: 92.927%, Rank #7: 93.659%, Rank #8: 95.122%, Rank #9: 95.122%, Rank #10: 95.366%, mAP:63.252%
Epoch: [78/100]| Loss: 0.12061



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.659%, Rank #2: 81.707%, Rank #3: 85.610%, Rank #4: 88.537%, Rank #5: 90.732%, Rank #6: 92.683%, Rank #7: 92.927%, Rank #8: 94.878%, Rank #9: 94.878%, Rank #10: 95.366%, mAP:63.290%
Epoch: [79/100]| Loss: 0.17559



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.610%, Rank #2: 83.902%, Rank #3: 86.341%, Rank #4: 87.805%, Rank #5: 90.244%, Rank #6: 91.951%, Rank #7: 93.415%, Rank #8: 94.146%, Rank #9: 94.634%, Rank #10: 95.122%, mAP:63.476%
Epoch: [80/100]| Loss: 0.10710



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.146%, Rank #2: 80.244%, Rank #3: 87.317%, Rank #4: 89.756%, Rank #5: 91.707%, Rank #6: 92.439%, Rank #7: 93.415%, Rank #8: 94.146%, Rank #9: 94.634%, Rank #10: 95.122%, mAP:62.878%
Epoch: [81/100]| Loss: 0.14361



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 71.951%, Rank #2: 81.220%, Rank #3: 84.146%, Rank #4: 87.317%, Rank #5: 88.293%, Rank #6: 90.244%, Rank #7: 90.732%, Rank #8: 91.707%, Rank #9: 92.439%, Rank #10: 93.659%, mAP:60.379%
Epoch: [82/100]| Loss: 0.13412



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 70.244%, Rank #2: 78.049%, Rank #3: 84.390%, Rank #4: 87.073%, Rank #5: 89.512%, Rank #6: 91.707%, Rank #7: 92.683%, Rank #8: 93.659%, Rank #9: 94.390%, Rank #10: 94.390%, mAP:60.037%
Epoch: [83/100]| Loss: 0.13116



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.927%, Rank #2: 79.024%, Rank #3: 84.878%, Rank #4: 87.805%, Rank #5: 90.488%, Rank #6: 91.463%, Rank #7: 91.951%, Rank #8: 92.195%, Rank #9: 93.171%, Rank #10: 93.659%, mAP:60.939%
Epoch: [84/100]| Loss: 0.15063



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 71.707%, Rank #2: 80.000%, Rank #3: 86.098%, Rank #4: 89.024%, Rank #5: 91.463%, Rank #6: 92.927%, Rank #7: 93.659%, Rank #8: 94.146%, Rank #9: 94.878%, Rank #10: 95.366%, mAP:62.242%
Epoch: [85/100]| Loss: 0.11417



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.146%, Rank #2: 81.951%, Rank #3: 88.293%, Rank #4: 90.244%, Rank #5: 91.951%, Rank #6: 92.439%, Rank #7: 93.171%, Rank #8: 93.902%, Rank #9: 94.878%, Rank #10: 95.366%, mAP:63.829%
Epoch: [86/100]| Loss: 0.15425



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 77.073%, Rank #2: 82.439%, Rank #3: 86.829%, Rank #4: 89.512%, Rank #5: 91.220%, Rank #6: 93.171%, Rank #7: 93.902%, Rank #8: 94.146%, Rank #9: 94.878%, Rank #10: 95.854%, mAP:63.151%
Epoch: [87/100]| Loss: 0.19016



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 75.854%, Rank #2: 83.415%, Rank #3: 86.585%, Rank #4: 90.488%, Rank #5: 91.463%, Rank #6: 92.439%, Rank #7: 93.659%, Rank #8: 94.146%, Rank #9: 94.878%, Rank #10: 95.122%, mAP:61.836%
Epoch: [88/100]| Loss: 0.15170



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.146%, Rank #2: 80.976%, Rank #3: 86.829%, Rank #4: 88.537%, Rank #5: 89.756%, Rank #6: 90.976%, Rank #7: 91.951%, Rank #8: 93.902%, Rank #9: 95.122%, Rank #10: 95.610%, mAP:60.904%
Epoch: [89/100]| Loss: 0.13383



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.634%, Rank #2: 82.439%, Rank #3: 87.073%, Rank #4: 88.780%, Rank #5: 90.732%, Rank #6: 92.195%, Rank #7: 94.146%, Rank #8: 95.366%, Rank #9: 95.610%, Rank #10: 95.854%, mAP:62.050%
Epoch: [90/100]| Loss: 0.12890



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.171%, Rank #2: 82.683%, Rank #3: 86.098%, Rank #4: 88.537%, Rank #5: 90.732%, Rank #6: 91.951%, Rank #7: 93.171%, Rank #8: 93.659%, Rank #9: 93.902%, Rank #10: 94.878%, mAP:62.217%
Epoch: [91/100]| Loss: 0.14347



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.927%, Rank #2: 81.707%, Rank #3: 84.390%, Rank #4: 86.829%, Rank #5: 90.488%, Rank #6: 92.195%, Rank #7: 92.683%, Rank #8: 93.415%, Rank #9: 93.659%, Rank #10: 94.390%, mAP:62.074%
Epoch: [92/100]| Loss: 0.11680



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 70.488%, Rank #2: 79.512%, Rank #3: 85.366%, Rank #4: 89.024%, Rank #5: 90.488%, Rank #6: 91.220%, Rank #7: 92.927%, Rank #8: 93.659%, Rank #9: 94.146%, Rank #10: 94.390%, mAP:60.670%
Epoch: [93/100]| Loss: 0.14662



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.683%, Rank #2: 80.976%, Rank #3: 85.854%, Rank #4: 88.293%, Rank #5: 90.488%, Rank #6: 91.220%, Rank #7: 93.171%, Rank #8: 94.390%, Rank #9: 95.366%, Rank #10: 95.854%, mAP:60.030%
Epoch: [94/100]| Loss: 0.13867



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.195%, Rank #2: 83.415%, Rank #3: 89.024%, Rank #4: 90.976%, Rank #5: 92.195%, Rank #6: 93.415%, Rank #7: 95.610%, Rank #8: 95.854%, Rank #9: 96.098%, Rank #10: 96.098%, mAP:61.597%
Epoch: [95/100]| Loss: 0.19110



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 72.195%, Rank #2: 82.683%, Rank #3: 87.317%, Rank #4: 90.976%, Rank #5: 92.195%, Rank #6: 93.659%, Rank #7: 94.878%, Rank #8: 96.098%, Rank #9: 96.098%, Rank #10: 96.585%, mAP:62.345%
Epoch: [96/100]| Loss: 0.14219



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 73.659%, Rank #2: 83.659%, Rank #3: 87.805%, Rank #4: 89.756%, Rank #5: 91.220%, Rank #6: 93.171%, Rank #7: 93.902%, Rank #8: 94.634%, Rank #9: 95.854%, Rank #10: 96.098%, mAP:63.819%
Epoch: [97/100]| Loss: 0.07859



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.390%, Rank #2: 84.146%, Rank #3: 88.293%, Rank #4: 90.732%, Rank #5: 92.195%, Rank #6: 93.659%, Rank #7: 94.146%, Rank #8: 94.878%, Rank #9: 95.610%, Rank #10: 96.098%, mAP:63.998%
Epoch: [98/100]| Loss: 0.08823



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 76.585%, Rank #2: 83.659%, Rank #3: 87.073%, Rank #4: 89.512%, Rank #5: 92.439%, Rank #6: 93.659%, Rank #7: 94.390%, Rank #8: 95.610%, Rank #9: 97.317%, Rank #10: 97.317%, mAP:64.662%
Epoch: [99/100]| Loss: 0.12396



Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 74.634%, Rank #2: 85.366%, Rank #3: 88.537%, Rank #4: 90.732%, Rank #5: 92.927%, Rank #6: 93.902%, Rank #7: 94.634%, Rank #8: 95.366%, Rank #9: 96.341%, Rank #10: 97.073%, mAP:65.033%
Epoch: [100/100]| Loss: 0.14823


 ## Performance Evaluation

In [20]:
# Initialize the GNN model with the same architecture and parameters
m1 = GraphAttentionNetwork(in_channels=100, temporal_dim=128, hidden_channels=256, out_channels=256).to(device)
# Load pre-trained model weights
models_folder = os.path.join('./output/models/')
m1.load_state_dict(torch.load(f'{models_folder}/best_GNN_state_r1.bin', map_location=device))

# Evaluate and print the performance of the model with the respective evaluation metric
print('Model w/ best R1:')
testing_step(m1, test_dataloader)

Model w/ best R1:

Estrazione feature della Gallery (410 grafi)...

Valutazione Probes in corso...


Total samples: 410, Rank #1: 77.073%, Rank #2: 82.439%, Rank #3: 86.829%, Rank #4: 89.512%, Rank #5: 91.220%, Rank #6: 93.171%, Rank #7: 93.902%, Rank #8: 94.146%, Rank #9: 94.878%, Rank #10: 95.854%, mAP:63.151%


(np.float64(0.7707317073170732),
 np.float64(0.9121951219512195),
 np.float64(0.9585365853658536),
 np.float64(0.6315126616995861))

Adding parallel temporal convolutions increased computational complexity but proved that the model's bottleneck is no longer the feature extraction or spatial routing mechanics. Since we have thoroughly exhausted GNN hyperparameter tuning, loss function margins, learning rate schedulers, attention heads, and temporal CNN receptive fields without moving past ~82.7%, the limitation is highly likely in the data representation itself.